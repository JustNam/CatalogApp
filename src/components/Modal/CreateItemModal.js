import React, { Component } from 'react';
import { Modal, Button, Form } from 'react-bootstrap';

class CreateItemModal extends Component {
  render() {
    const { show, title, description, handleTitleChange, handleDescriptionChange, handleClose, handleSubmit } = this.props;
    return (
      <Modal animation={false} show={show} onHide={handleClose}>
        <Modal.Header closeButton>
          <Modal.Title>Create new item</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Form onSubmit={handleSubmit}>
            <Form.Group controlId="formBasicTitle">
              <Form.Label>Title</Form.Label>
              <Form.Control
                type="input"
                name="title"
                onChange={handleTitleChange}
                value={title}
              />
              <Form.Text className="text-muted">
                Title only contain letters, numbers and spaces. Minimum length is 6 characters,
                maximum is 30.
              </Form.Text>
            </Form.Group>

            <Form.Group controlId="formBasicPassword">
              <Form.Label>Description</Form.Label>
              <Form.Control
                type="textarea"
                name="description"
                onChange={handleDescriptionChange}
                value={description}
              />
            </Form.Group>
            <Button type="submit">Create</Button>
          </Form>
        </Modal.Body>
      </Modal>
    );
  }
}

export default CreateItemModal;
