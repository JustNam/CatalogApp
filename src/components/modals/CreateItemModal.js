import React, { Component } from 'react';
import { Modal, Button, Form } from 'react-bootstrap';
import { connect } from 'react-redux';
import { createItemInCategory } from 'actions/item';
import { validateItemTitle } from 'utilities/validate';

class CreateItemModal extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isValidRequest: false,
      title: {
        value: '',
        isValid: false,
        errorMessages: [],
      },
      description: {
        value: '',
      },
      // Attribute for <input/> element
      titleInputProps: {
        isValid: null,
        isInvalid: null,
      },
    };
  }

  // Update UI of title input
  updateTitleInputProps = (validity) => {
    this.setState({
      titleInputProps: {
        isValid: validity,
        isInvalid: !validity,
      },
    });
  };

  resetTitleInputProps = () => {
    this.setState({
      titleInputProps: {
        isValid: '',
        isInvalid: '',
      },
    });
  };

  // Validate title of item and then save it to state
  validateData = (title) => {
    const validation = validateItemTitle(title);
    this.setState({
      isValidRequest: validation.isValid,
      title: {
        ...title,
        ...validation,
      },
    });
    this.updateTitleInputProps(validation.isValid);
  };

  // Update title from input and trigger validadtions
  handleTitleChange = (e) => {
    const value = e.target.value;
    if (value === '') {
      this.setState({
        title: {
          value: '',
          isValid: false,
          errorMessages: [],
        },
      });
      this.resetTitleInputProps();
    } else {
      // Can not use this assignment when the input is empty
      const newTitle = {
        ...this.state.title,
        value,
      };
      this.validateData(newTitle);
    }
  };

  handleDescriptionChange = (e) => {
    this.setState({ description: { value: e.target.value } });
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const {
      title,
      description,
      isValidRequest,
    } = this.state;
    // Validate by front-end
    if (isValidRequest) {
      // Submit new item information
      this.props
        .createItemInCategory(this.props.currentCategoryId, {
          title: title.value.trim(),
          description: description.value,
        })
        .then((response) => {
          // Validate by back-end
          if (response.payload.error) {
            this.setState({
              title: {
                ...title,
                errorMessages: ['The given title already exists'],
              },
            });
            this.updateTitleInputProps(false);
          } else {
            this.props.showSuccessModal();
            this.props.handleClose();
          }
        });
    }
  };

  render() {
    const { title, titleInputProps, description } = this.state;
    return (
      <Modal
        animation={false}
        show={this.props.show}
        onHide={this.props.handleClose}
      >
        <Modal.Header closeButton>
          <Modal.Title>Create new item</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Form onSubmit={this.handleSubmit}>
            <Form.Group controlId="formBasicTitle">
              <Form.Label>Title</Form.Label>
              <Form.Control
                name="title"
                required
                onChange={this.handleTitleChange}
                value={title.value}
                {...titleInputProps}
              />
              <Form.Text className="text-muted">
                Title only contain letters, numbers and spaces. Maximum length
                is 30 characters.
              </Form.Text>
              {/* Show error detais */}
              <Form.Control.Feedback type="invalid">
                {title.errorMessages.map((errorMessage, index) => (
                  <div key={index}>{errorMessage}</div>
                ))}
              </Form.Control.Feedback>
            </Form.Group>

            <Form.Group controlId="formBasicPassword">
              <Form.Label>Description</Form.Label>
              <Form.Control
                as="textarea"
                name="description"
                onChange={this.handleDescriptionChange}
                value={description.value}
              />
            </Form.Group>
            <Button type="submit">Create</Button>

          </Form>
        </Modal.Body>
      </Modal>
    );
  }
}

const mapDispatchToProp = {
  createItemInCategory,
};

export default connect(
  null,
  mapDispatchToProp
)(CreateItemModal);
export { CreateItemModal };