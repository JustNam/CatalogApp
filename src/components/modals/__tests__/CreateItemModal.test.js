import React from 'react';
import { shallow } from 'enzyme';
import { CreateItemModal } from '../CreateItemModal';

describe('CreateItemModal with valid data', () => {
  let wrapper;
  let props;
  let instance;
  const setup = () => {
    props = {
      isValidRequest: false,
      title: {
        value: '',
        isValid: false,
        errorMessages: [],
      },
      description: {
        value: '',
      },
      titleInputProps: {
        isValid: null,
        isInvalid: null,
      },
      createItemInCategory: jest.fn(() => Promise.resolve({
        payload: {
          data: '',
        },
      })),
      showSuccessModal: jest.fn(() => {}),
      handleClose: jest.fn(() => {}),
    };
  };
  beforeEach(() => {
    setup();
    wrapper = shallow(<CreateItemModal {...props} />);
    instance = wrapper.instance();
  });
  it('It should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });
  it('It should change description in state when description input is changed', () => {
    expect(wrapper.state().isValidRequest).toBe(false);
  });
  it('It should change username in state when username input is changed', () => {
    instance.handleTitleChange({
      target: {
        value: 'test',
      },
    });
    expect(wrapper.state().title.value).toBe('test');
    expect(wrapper.state().titleInputProps.isInvalid).toBe(false);

    instance.handleTitleChange({
      target: {
        value: '',
      },
    });
    expect(wrapper.state().title.value).toBe('');
  });
  it('It should change description in state when description input is changed', () => {
    instance.handleDescriptionChange({
      target: {
        value: 'test',
      },
    });
    expect(wrapper.state().description.value).toBe('test');
  });
  it('It should show success modal if the data is valid', async () => {
    wrapper.setState({
      isValidRequest: true,
    });
    await instance.handleSubmit({
      preventDefault: () => {},
    });
    expect(props.showSuccessModal).toBeCalled();
  });
});

describe('CreateItemModal with valid data', () => {
  let wrapper;
  let props;
  let instance;
  const setup = () => {
    props = {
      isValidRequest: false,
      title: {
        value: '',
        isValid: false,
        errorMessages: [],
      },
      description: {
        value: '',
      },
      titleInputProps: {
        isValid: null,
        isInvalid: null,
      },
      createItemInCategory: jest.fn(() => Promise.resolve({
        payload: {
          error: '1',
        },
      })),
      showSuccessModal: jest.fn(() => {}),
      handleClose: jest.fn(() => {}),
    };
  };
  beforeEach(() => {
    setup();
    wrapper = shallow(<CreateItemModal {...props} />);
    instance = wrapper.instance();
  });
  it('It should show error if the data is considered invalid in back end', async () => {
    wrapper.setState({
      isValidRequest: true,
    });
    await instance.handleSubmit({
      preventDefault: () => {},
    });
    expect(wrapper.state().titleInputProps.isValid).toBe(false);
  });

  it('It should not call API if the data is considered invalid in front end', async () => {
    wrapper.setState({
      isValidRequest: false,
    });
    await instance.handleSubmit({
      preventDefault: () => {},
    });
    expect(props.createItemInCategory).not.toHaveBeenCalled();
  });
});