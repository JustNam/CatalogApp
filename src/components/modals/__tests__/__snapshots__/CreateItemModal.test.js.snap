// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateItemModal with valid data It should render correctly 1`] = `ShallowWrapper {}`;
